### PR TITLE
Simplify group_images_by_shape

### PR DESCRIPTION
**Pull request by guarin**
# What does this PR do?

* Simplifies `group_images_by_shape` and moves all logic into one function
* Allow passing `None` as `paired_input`, required for https://github.com/huggingface/transformers/pull/48066
* Remove need for `is_nested` in `reorder_images` as the index keeps track of it
* Updated models that referred to `key` as `shape` which is not always true

Follow-up from https://github.com/huggingface/transformers/pull/47964#discussion_r3782734308

Ran slow image processing tests for all models locally.

CC @zucchini-nlp

### CODE REVIEW

File: src/transformers/image_transforms.py [hunk 1]
```diff
@@ -1006,83 +925,100 @@ def group_images_by_shape(
...
-    # Stack images with the same shape. Groups holding a single image are unsqueezed instead, as stacking would copy
-    # the image for no reason.
+    # Group images and paired inputs
+    grouped_images = {}
+    paired_inputs_groups = [{} for _ in paired_inputs]
+    positions = {}
+    for sublist_index, sublist in enumerate(nested_images):
+        for image_index, image in enumerate(sublist):
+            position = (sublist_index, image_index)
+            key = position if disable_grouping else image.shape[-2:]
```
Comment (guarin) [inline src/transformers/image_transforms.py]: Strictly speaking `torch.stack` will only pass if `key = image.shape` (including channel dim) but we assume that channel dims are the same.

VitMatte relies on this as it groups once images and once trimaps with a single channel and both groupings must end up with the same keys.
